tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/common/BatchConfiguration.java

```diff
@@ -69,63 +71,48 @@ public static BatchConfiguration computeChunkSizes(TornadoExecutionContext conte
 
         for (Object o : inputObjects) {
             if (o.getClass().isArray()) {
                 Class<?> componentType = o.getClass().getComponentType();
-                dataTypeSize = findDataTypeSize(componentType);
+                DataTypeSize dataTypeSize = findDataTypeSize(componentType);
                 if (dataTypeSize == null) {
                     throw new TornadoRuntimeException("[UNSUPPORTED] Data type not supported for processing in batches");
                 }
                 long size = Array.getLength(o);
                 totalSize = size * dataTypeSize.getSize();
 
-                classObjects.add(componentType);
+                elementSizes.add(dataTypeSize.getSize());
                 inputSizes.add(totalSize);
-                if (classObjects.size() > 1) {
-                    throw new TornadoRuntimeException("[UNSUPPORTED] Input objects with different data types not currently supported");
-                }
-                if (inputSizes.size() > 1) {
-                    throw new TornadoRuntimeException("[UNSUPPORTED] Input objects with different sizes not currently supported");
-                }
-            } else if (o instanceof IntArray) {
-                totalSize = ((IntArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(int.class);
-            } else if (o instanceof FloatArray) {
-                totalSize = ((FloatArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(float.class);
-            } else if (o instanceof DoubleArray) {
-                totalSize = ((DoubleArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(double.class);
-            } else if (o instanceof LongArray) {
-                totalSize = ((LongArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(long.class);
-            } else if (o instanceof ShortArray) {
-                totalSize = ((ShortArray) o).getNumBytesWithoutHeader();
+            } else if (o instanceof TornadoNativeArray tornadoNativeArray) {
+                totalSize = tornadoNativeArray.getNumBytesWithoutHeader();
                 inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(short.class);
-            } else if (o instanceof ByteArray) {
-                totalSize = ((ByteArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(byte.class);
-            } else if (o instanceof CharArray) {
-                totalSize = ((CharArray) o).getNumBytesWithoutHeader();
-                inputSizes.add(totalSize);
-                dataTypeSize = findDataTypeSize(char.class);
+                byte elementSize = switch (tornadoNativeArray) {
```

Comment:
This can be further simplified, e.g. I saw that the arrays all have the `SegmentElementSize` annotation on their type. Would it make sense to use it here?

Reply:
We only want to use the annotation part in the compiler side, not in the runtime. We try to avoid reflection as much as possible.